Add clock plugin with basic functionality

xela-95 opened the pull request:
This PR adds to the gz-sim-yarp-plugins a clock plugin, with the only basic functionality of writing the current simulation time on the yarp `/clock` port.

Closes #60 , #73

xela-95 committed:
Add clock plugin folder with basic implementation to get sim time and print it

xela-95 committed:
Add world file to clock tutorial

xela-95 committed:
Write sim time on /clock buffered port

traversaro commented on plugins/clock/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+project(clock)
```
```suggestion
```

xela-95 replied:
@traversaro we can remove this because it is a low level plugin that is not referenced in other CMakeLists?

traversaro replied:
Not really. The main point is that `project` only make sense inside the root CMakeLists.txt of a project, and typically it contain project-specific metadata like the version and the programming languages used (see https://cmake.org/cmake/help/latest/command/project.html). In some specific cases, it could make sense to have it in a subdirectory, but in general it is only useless code that can create confusion.

traversaro commented on plugins/clock/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+  PRIVATE gz-plugin${GZ_PLUGIN_VER}::gz-plugin${GZ_PLUGIN_VER}
+  PRIVATE gz-sim${GZ_SIM_VER}::gz-sim${GZ_SIM_VER}
+  PRIVATE
+    ${YARP_LIBRARIES}
```
```suggestion
```

If you need more `YARP` targets, let's list them explicitly (probably `YARP::YARP_dev` and/or `YARP::YARP_sig` are necessary).

xela-95 replied:
@traversaro could you point me towards some reference in which to better understand this? I've seen in Yarp documentation the usage of `${YARP_LIBRARIES}`: https://yarp.it/latest/group__add__a__plugin.html

traversaro replied:
Good point, in general this would need to be updated. The point is that you never actually know what the `${YARP_LIBRARIES}` variable will contain, as it depends on how `find_package(YARP)` has been called (with which arguments). So in so-called "Modern CMake" it is better to always use targets, that clearly identify a single library. In your case, just try to put `YARP::YARP_dev` in place of `${YARP_LIBRARIES}`, and if there are compilation error we can look into that.

xela-95 replied:
Solve in 8f6967faa26c75debd5b0443e3d0180ca9f9d55e

Putting YARP::YARP_dev in place of ${YARP_LIBRARIES} worked.

xela-95 committed:
Update plugins/clock/CMakeLists.txt

Remove useless project statement

Co-authored-by: Silvio Traversaro <silvio@traversaro.it>

xela-95 committed:
Update YARP library target in CMakeLists.txt

xela-95 committed:
Remove debug logging

xela-95 committed:
Add unit test for clock plugin

xela-95 committed:
Update unit test

xela-95 commented:
@traversaro I tried to develop a first basic unit test for the clock plugin, here: 

https://github.com/robotology/gz-sim-yarp-plugins/blob/4a06b08fa297f69f3959402f0ecd1d0dd23963c7/tests/clock/ClockTest.cc

The unit test instantiates a gazebo server fixture and reads the `/clock` port after a certain number of simulation steps have been ran. Then it checks whether the elapsed simulation time received on the port is the one expected (number of timesteps times delta T).

Next I will try to add test to check other expected behaviors, like the pause and reset of simulation.

xela-95 committed:
WIP Add mechanism to avoid sending timestamp on clock port

xela-95 committed:
Use paused data member to check if simulation is paused instead of comparing timestamps

traversaro commented on plugins/clock/Clock.cc:
```diff
@@ -20,35 +21,70 @@ using yarp::os::Network;
 
     ~Clock()
     {
+        std::cout << "===========> destructor" << std::endl;
```
When debugging, always use `std::cerr`, `std::cout` may be buffered (even if in theory with `std::endl` you flush it).

xela-95 replied:
Ok thank you!

xela-95 committed:
Manage plugin reset

In this way it is possible to avoid desctruction and re-construction of the plugin object with consequent closing of  the yarp port.

xela-95 commented:
In commit 5acde4f08c7066d4afbf20918ab7b7df9a16cac9 I've implemented the `ISystemReset` interface. In this way, every time the user requests a simulation reset through Gazebo UI this method is called instead of destructing and re-constructing the plugin with the consequent closing of the yarp port. Now for a subscriber to the port the reset is transparent and it will just see the simulation time starting back from zero.

Unfortunately I've been not able to implement a unit test since the Server object does not expose a `Reset()` method that I can use to inject such event.

Some issues and PRs talking about reset: 
- https://github.com/gazebosim/gz-sim/issues/1107
- https://github.com/gazebosim/gz-sim/issues/203#issuecomment-642138544

xela-95 committed:
Add tutorial README

traversaro commented on plugins/clock/Clock.cc:
```diff
@@ -0,0 +1,91 @@
+    {
+        if (_info.paused)
+        {
+            // yDebug() << "Simulation pause, skipping clock update";
```
```suggestion
            // yDebug() << "Simulation pause, skipping clock update";
```
Similarly, if a message is commented, we can just remove it from the code to keep code simple.

traversaro reviewed (approved):
Just a small comment regarding unnecessary debug messages. Beside that, I think it is good to go, remember to squash commits!

xela-95 commented on plugins/clock/Clock.cc:
```diff
@@ -0,0 +1,91 @@
+        : m_portName("/clock")
+        , m_initialized(false)
+    {
+        yDebug() << "Clock plugin Constructor invoked";
```
Oh sure, I thought these prints will be enabled by manually setting a higher level of verbosity, but actually I have seen them while debugging the code without doing anything. I will remove them.

xela-95 committed:
Update plugins/clock/Clock.cc

Co-authored-by: Silvio Traversaro <silvio@traversaro.it>

xela-95 committed:
Update logs
